Security/TwoFactor/CsrfProtection/CsrfTokenValidator.php

```diff
@@ -0,0 +1,48 @@
+<?php
+
+namespace Scheb\TwoFactorBundle\Security\TwoFactor\CsrfProtection;
+
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\Security\Csrf\CsrfToken;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+
+class CsrfTokenValidator
+{
+    /**
+     * @var string
+     */
+    private $firewallName;
+
+    /**
+     * @var array
+     */
+    private $options;
+
+    public function __construct(string $firewallName, array $options)
```

Comment:
I'd change the implementation of this to simplify things a bit:
- Inject the CSRF manager as an argument, so that it's not possible to inject anything else but `CsrfTokenManagerInterface`. We can then remove all the `instanceof` checks because it cannot by anything else. And in case someone misconfigures, they'd recognize as soon as the container tries to build - and fails.
- The bundle should come with a `NullCsrfTokenManager`, which is always returning true. This one should be injected by default, when there's no specific CSRF manager configured by the firewall. I've implemented the same concept for the optional `BackupCodeManager ` and `TrustedDeviceManager`, makes things so much simpler because the calling method doesn't need to worry about optional features. Would simplify TwoFactorListener.php#188 to just a call to `CsrfTokenValidator`.

Reply:
This feels like a better approach indeed.